File: examples/python/h2_vqe.py
"""使用 ArcQML Python API 求解 H2/STO-3G 基态能量。"""

from __future__ import annotations

import json
import math
from pathlib import Path

import arcqml


NUM_QUBITS = 4
ACTIVE_ELECTRONS = 2
LAYERS = 6
STEPS = 100
EXACT_GROUND_ENERGY = -1.1361894542078266
HARTREE_FOCK_ENERGY = -1.1173490350562805
CHEMICAL_ACCURACY = 1.6e-3
DATA_PATH = Path(__file__).resolve().parents[1] / "data" / "h2_hamiltonian.json"


def load_hamiltonian(path: Path) -> arcqml.PauliSum:
    """从 examples/data 的公共 JSON 定义构造 PauliSum。"""
    payload = json.loads(path.read_text(encoding="utf-8"))
    hamiltonian = arcqml.PauliSum(num_qubits=payload["num_qubits"])
    for term in payload["terms"]:
        coefficient = term["coefficient"]
        operations = term["paulis"]
        if not operations:
            hamiltonian.add_identity(coefficient=coefficient)
            continue
        hamiltonian.add_term(
            paulis="".join(operation["pauli"] for operation in operations),
            qubits=[operation["qubit"] for operation in operations],
            coefficient=coefficient,
        )
    return hamiltonian


def build_ansatz() -> arcqml.Circuit:
    """构造 Hartree-Fock 初态和 6 层硬件高效 ansatz。"""
    circuit = arcqml.Circuit(num_qubits=NUM_QUBITS)
    for qubit in range(ACTIVE_ELECTRONS):
        circuit.x(qubit=qubit)

    parameter_index = 0
    for _layer in range(LAYERS):
        for qubit in range(NUM_QUBITS):
            ry = 0.04 * math.sin(0.37 * (parameter_index + 1))
            parameter_index += 1
            rz = 0.04 * math.sin(0.37 * (parameter_index + 1))
            parameter_index += 1
            circuit.ry(angle=ry, qubit=qubit)
            circuit.rz(angle=rz, qubit=qubit)
        for control in range(NUM_QUBITS):
            circuit.cnot(control=control, target=(control + 1) % NUM_QUBITS)
    return circuit


def evaluate(
    simulator: arcqml.StateVectorSimulator,
    circuit: arcqml.Circuit,
    hamiltonian: arcqml.PauliSum,
) -> float:
    """无梯度计算当前能量。"""
    with arcqml.no_grad():
        return float(simulator.run(circuit=circuit, observable=hamiltonian).item())


def main() -> None:
    hamiltonian = load_hamiltonian(path=DATA_PATH)
    circuit = build_ansatz()
    simulator = arcqml.StateVectorSimulator(num_qubits=NUM_QUBITS)
    optimizer = arcqml.Adam(learning_rate=0.05)

    print(f"H2/STO-3G, Jordan-Wigner, {NUM_QUBITS} qubits")
    print(f"Hamiltonian terms : {hamiltonian.num_terms}")
    print(f"Trainable params  : {circuit.num_parameters}")
    print(f"Hartree-Fock      : {HARTREE_FOCK_ENERGY:.12f} Ha")
    print(f"Exact ground      : {EXACT_GROUND_ENERGY:.12f} Ha")

    for step in range(1, STEPS + 1):
        energy = simulator.run(circuit=circuit, observable=hamiltonian)
        energy.backward()
        optimizer.step(circuit=circuit)
        optimizer.zero_grad(circuit=circuit)

        if step == 1 or step % 10 == 0:
            current = evaluate(
                simulator=simulator,
                circuit=circuit,
                hamiltonian=hamiltonian,
            )
            error = abs(current - EXACT_GROUND_ENERGY)
            print(f"step {step:>3}: energy={current:.12f} Ha, |error|={error:.3e} Ha")

    final_energy = evaluate(
        simulator=simulator,
        circuit=circuit,
        hamiltonian=hamiltonian,
    )
    error = abs(final_energy - EXACT_GROUND_ENERGY)
    reached = "reached" if error < CHEMICAL_ACCURACY else "not reached"
    print(f"\nVQE energy        : {final_energy:.12f} Ha")
    print(f"Absolute error    : {error:.6e} Ha")
    print(f"Chemical accuracy: {reached} (threshold {CHEMICAL_ACCURACY:.1e} Ha)")


if __name__ == "__main__":
    main()

File: examples/rust/h2_vqe.rs
use arcqml::prelude::*;
use arcqml_visualization::write_svg;

type AppResult<T> = std::result::Result<T, Box<dyn std::error::Error>>;

const NUM_QUBITS: usize = 4;
const ACTIVE_ELECTRONS: usize = 2;
const LAYERS: usize = 6;
const STEPS: usize = 100;
const EXACT_GROUND_ENERGY: f64 = -1.136_189_454_207_826_6;
const HARTREE_FOCK_ENERGY: f64 = -1.117_349_035_056_280_5;
const CHEMICAL_ACCURACY: f64 = 1.6e-3;

fn h2_ansatz() -> AppResult<Circuit> {
    let mut circuit = Circuit::new(NUM_QUBITS)?;

    // Jordan–Wigner 编码下的 Hartree–Fock 初态：q0、q1 被占据。
    for qubit in 0..ACTIVE_ELECTRONS {
        circuit.x(qubit)?;
    }

    let mut parameter_index = 0usize;
    for _ in 0..LAYERS {
        for qubit in 0..NUM_QUBITS {
            let ry = 0.04 * (0.37 * (parameter_index + 1) as f64).sin();
            parameter_index += 1;
            let rz = 0.04 * (0.37 * (parameter_index + 1) as f64).sin();
            parameter_index += 1;
            circuit.ry(ry, qubit)?;
            circuit.rz(rz, qubit)?;
        }
        for control in 0..NUM_QUBITS {
            circuit.cnot(control, (control + 1) % NUM_QUBITS)?;
        }
    }
    Ok(circuit)
}

fn energy(
    simulator: &StateVectorSimulator,
    circuit: &Circuit,
    hamiltonian: &SparsePauliOp,
) -> AppResult<f64> {
    Ok(simulator.run(circuit, hamiltonian)?.value()?)
}

fn main() -> AppResult<()> {
    let hamiltonian = SparsePauliOp::from_json(include_str!(
        "../data/h2_hamiltonian.json"
    ))?;
    let circuit = h2_ansatz()?;
    let simulator = StateVectorSimulator::new(NUM_QUBITS)?;
    let mut optimizer = Adam::new(0.05, 0.9, 0.999, 1e-8, 0.0)?;

    println!("H2/STO-3G, Jordan-Wigner, {} qubits", NUM_QUBITS);
    println!("Hamiltonian terms : {}", hamiltonian.len());
    println!("Trainable params  : {}", circuit.num_parameters());
    println!("Hartree-Fock      : {HARTREE_FOCK_ENERGY:.12} Ha");
    println!("Exact ground      : {EXACT_GROUND_ENERGY:.12} Ha");

    for step in 1..=STEPS {
        let energy_tensor = simulator.run(&circuit, &hamiltonian)?;
        energy_tensor.backward()?;
        optimizer.step(circuit.parameters())?;
        optimizer.zero_grad(circuit.parameters());

        if step == 1 || step % 10 == 0 {
            let current = energy(&simulator, &circuit, &hamiltonian)?;
            println!(
                "step {step:>3}: energy = {current:.12} Ha, |error| = {:.3e} Ha",
                (current - EXACT_GROUND_ENERGY).abs()
            );
        }
    }

    write_svg(&circuit, "./examples/rust/h2_vqe_circuit.svg");

    let final_energy = energy(&simulator, &circuit, &hamiltonian)?;
    let absolute_error = (final_energy - EXACT_GROUND_ENERGY).abs();
    println!("\nVQE energy        : {final_energy:.12} Ha");
    println!("Absolute error    : {absolute_error:.6e} Ha");
    println!(
        "Chemical accuracy: {} (threshold {:.1e} Ha)",
        if absolute_error < CHEMICAL_ACCURACY {
            "reached"
        } else {
            "not reached"
        },
        CHEMICAL_ACCURACY
    );
    Ok(())
}
